src/attachment.rs

```diff
@@ -0,0 +1,404 @@
+use std::{borrow::Cow, cell::UnsafeCell, collections::HashMap};
+
+use libc::c_void;
+
+use crate::{z_bytes_null, z_bytes_t};
+
+use zenoh::sample::Attachment;
+
+/// The body of a loop over an attachment's key-value pairs.
+///
+/// `key` and `value` are loaned to the body for the duration of a single call.
+/// `context` is passed transparently through the iteration driver.
+///
+/// Returning `0` is treated as `continue`.
+/// Returning any other value is treated as `break`.
+pub type z_attachment_iter_body_t =
+    extern "C" fn(key: z_bytes_t, value: z_bytes_t, context: *mut c_void) -> i8;
+
+/// The driver of a loop over an attachment's key-value pairs.
+///
+/// This function is expected to call `loop_body` once for each key-value pair
+/// within `iterator`, passing `context`, and returning any non-zero value immediately (breaking iteration).
+pub type z_attachment_iter_driver_t = extern "C" fn(
+    iterator: *const c_void,
+    loop_body: z_attachment_iter_body_t,
+    context: *mut c_void,
+) -> i8;
+
+/// Returns the number of key-value pairs within the attachment.
+pub type z_attachment_len_t = extern "C" fn(*const c_void) -> usize;
+
+/// The v-table for an attachment.
+#[repr(C)]
+pub struct z_attachment_vtable_t {
+    /// See `z_attachment_iteration_driver_t`'s documentation.
+    iteration_driver: z_attachment_iter_driver_t,
+    /// Returns the number of key-value pairs within the attachment.
+    len: z_attachment_len_t,
+}
+
+/// Constructs a specific :c:type:`z_attachment_vtable_t`.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn z_attachment_vtable(
+    iteration_driver: z_attachment_iter_driver_t,
+    len: z_attachment_len_t,
+) -> z_attachment_vtable_t {
+    z_attachment_vtable_t {
+        iteration_driver,
+        len,
+    }
+}
+
+/// A v-table based map of byte slice to byte slice.
+///
+/// `vtable == NULL` marks the gravestone value, as this type is often optional.
+/// Users are encouraged to use `z_attachment_null` and `z_attachment_check` to interact.
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct z_attachment_t {
+    data: *const c_void,
+    vtable: Option<&'static z_attachment_vtable_t>,
+}
+
+/// Returns the gravestone value for `z_attachment_t`.
+#[no_mangle]
+pub extern "C" fn z_attachment_check(this: &z_attachment_t) -> bool {
+    this.vtable.is_some()
+}
+
+/// Returns the gravestone value for `z_attachment_t`.
+#[no_mangle]
+pub extern "C" fn z_attachment_null() -> z_attachment_t {
+    z_attachment_t {
+        data: core::ptr::null_mut(),
+        vtable: None,
+    }
+}
+
+/// Constructs a specific :c:type:`z_attachment_t`.
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub extern "C" fn z_attachment(
```

Comment:
This isn't really useful, since that's just a constructor for a structure that only has public fields